### PR TITLE
13 many target qubits automatic swapping

### PR DESCRIPTION
**Pull request by guskal01**

### CODE REVIEW

**Review by BAMF0 — CHANGES_REQUESTED** [review]
Looks good, but there should be swap tests in order to validate the functionality.

File: src/register.rs [hunk 1]
```diff
@@ -20,52 +20,65 @@ pub struct Register {
...
-        // otherwise returns 2 by 2 identity matrix
-        let get_matrix = |i| { if i == target { return op.matrix(); }
-            else { return operation::identity(0).matrix(); }
-        };
+        // Create a copy of state and permute the qubits according to perm
+        let mut permuted_state = self.state.clone();
+        for i in 0..permuted_state.len() {
+            let mut j: usize = 0;
+            for k in 0..self.size {
+                j |= ((i >> k) & 1) << perm[k];
```
Comment (simonengstrm) [inline src/register.rs:62]: I think this code should be commented more. It is very hard to read and understand what it does other than that it permutes the qubits according to perm.

File: src/register.rs [hunk 2]
```diff
@@ -20,52 +20,65 @@ pub struct Register {
...
     }
     /// Applys a quantum operation to the current state
     ///
     /// Input a state and an operation. Outputs the new state
     pub fn apply(&mut self, op: &Operation) -> &mut Self {
-        // Gets the target bit
-        let target = op.targets()[op.arity() - 1];
-        // Calculates the number of matrices in tensor product
-        let num_matrices = self.size + 1 - op.targets().len();
+        // Permutation which the qubits will be shuffled according to
```
Comment (simonengstrm) [inline src/register.rs:48]: Unfinished docs?

**Review by simonengstrm — CHANGES_REQUESTED** [review]
There are a couple of places where I would appreciate some more documentation. Other than this and some more tests with swapping it looks good.

**Review by BAMF0 — APPROVED** [review]
Looks good

**Comment by BAMF0** [comment]
I merge
 now

**Pull request merged**